docker/android/scripts/build-system.py

```diff
@@ -0,0 +1,207 @@
+    raise TypeError(f'got unexpected type of {type(item)}')
+
+
+def remove_soong_tests(path, args, *_):
+    print_verbose(f'removing soong tests from "{path}"', args.verbose)
+    with open(path) as file:
+        ast = android.soong.load(file)
+    # remove the test or benchmark scopes, IE, this with `cc_test`
+    # or those with `{name: "test"}`, etc.
+    ast.filter(lambda x: not (x.is_scope() and x.is_dev()))
+    # need to remove test and benchmark subdirs
+    test_names = ('test', 'benchmark')
+    subdirs = [i for i in ast if i.name == 'subdirs']
+    for sub in subdirs:
+        assert type(sub.expr) is android.soong.List
+        filter_list(sub.expr, test_names)
+    # remove gtest dependencies from regular targets.
+    for node in ast:
+        map = None
+        if not node.is_scope() and not node.expr.is_map():
+            continue
+        if node.is_scope():
+            map = node.map
+        else:
+            map = node.expr
+        test_names = ('libgtest', 'test-proto', 'starlarktest')
+        for key, value, *_ in map.recurse():
+            if value.value.is_list():
+                if key == 'testSrcs':
+                    value.value.clear()
+                else:
+                    filter_list(value, test_names)
+
+    with open(path, 'w') as file:
+        ast.dump(file)
+
+
+def remove_makefile_tests(path, args, *_):
+    print_verbose(f'removing makefile tests from "{path}"', args.verbose)
+    with open(path) as file:
+        makefile = android.make.load(file)
+    makefile.filter(lambda x: not x.is_dev())
+    with open(path, 'w') as file:
+        makefile.dump(file)
+
+
+def remove_tests(path, args, processor):
+    if os.path.exists(path + '.bak'):
+        restore(path, args)
+    elif not args.disable_backup:
+        backup(path, args)
+    processor(path, args)
+
+
+def stash(root):
+    git_glob = f'{root}/**/.git'
+    for path in glob.iglob(git_glob, recursive=True):
+        os.chdir(os.path.dirname(path))
+        subprocess.check_call(['git', 'stash'])
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    action_group = parser.add_mutually_exclusive_group(required=True)
+    action_group.add_argument(
+        '--backup',
+        help='backup build files',
+        action='store_true',
+    )
+    action_group.add_argument(
+        '--restore',
+        help='restore build files',
+        action='store_true',
+    )
+    action_group.add_argument(
+        '--remove-tests',
+        help='remove most tests from the build system.',
+        action='store_true',
+    )
+    action_group.add_argument(
+        '--stash',
+        help='stash all local changes.',
+        action='store_true',
+    )
+    parser.add_argument(
+        '--disable-backup',
+        help='disable automatic backup of build files during processing.',
+        action='store_false',
+    )
+    flags_group = parser.add_mutually_exclusive_group()
+    flags_group.add_argument(
+        '--soong-only',
```

Comment:
These are quite handy if in the future something breaks with our Makefile parser, but not our Soong parser, or the inverse, and tests do not need to be removed from the other. This might be a reality since the Makefiles are slowly being deprecated and we have an incomplete parser for them: only our Soong parser is correct.